### PR TITLE
Update Tools repo Sync to make use of patch files

### PR DESCRIPTION
**Pull request by chidozieononiwu**
For pull request triggered runs of the sync pipeline. Rather that copying over all the files in the Sync directory make use of a patch file to apply only the changes associated with the latest commits.

### CODE REVIEW

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 1]
```diff
@@ -14,17 +14,32 @@ steps:
...
+    - ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
+      - pwsh: |
+          git clone --branch ${{ parameters.BaseBranchName }} https://github.com/azure/${{ repo }}
+          $repoPath = "${{ repo }}/${{ parameters.DirectoryToSync }}"
+          $patchDir = "${{ repo }}/patchfiles"
 
-      displayName: Copy ${{ parameters.DirectoryToSync }} from azure-sdk-tools to ${{ repo }}
-      workingDirectory: $(System.DefaultWorkingDirectory)
+          pushd $(Build.SourcesDirectory)
+          git format-patch ${{ parameters.BaseBranchName }}...upstream/master ${{ parameters.DirectoryToSync }} -o $patchDir
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:36]: The patch should only be produced once from the tools repo not everytime from the repo we are trying to sync. We only want to apply the changes

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 2]
```diff
@@ -14,17 +14,32 @@ steps:
...
+          popd
+          pushd ${{ repo }}
+          Get-ChildItem $patchDir | % { git am $_.FullName }
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:39]: We need to makes sure we handle failures correctly here.

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 3]
```diff
@@ -14,17 +14,32 @@ steps:
...
+          git clean -xfd
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:40]: Why are you running a clean?

File: eng/pipelines/templates/steps/sync-directory.yml (hunk 2 above)
Comment (weshaggard) [reply eng/pipelines/templates/steps/sync-directory.yml:39]: git am will actually make commits but the other parts of this yml file below assume there is a dirty repo and it will try to add and make a commit and if there is no diff it will skip the push and pull request creation. We might need to make some changes to account for this new scenario.

File: eng/pipelines/templates/steps/sync-directory.yml (hunk 1 above)
Comment (chidozieononiwu) [reply eng/pipelines/templates/steps/sync-directory.yml:36]: Am actually producing the patch from the Tools Repo but storing the output in the repo bring synced to.

File: eng/pipelines/templates/steps/sync-directory.yml (hunk 3 above)
Comment (chidozieononiwu) [reply eng/pipelines/templates/steps/sync-directory.yml:40]: Because I dumped the patch files into the repo being synced.

Comment (weshaggard) [reply eng/pipelines/templates/steps/sync-directory.yml:40]: We should probably put them in another directory outside of the source.

File: eng/common/scripts/modules/ChangeLog-Operations.psm1 [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # Common Changelog Operations
-
+# This Files has been retired
```
Comment (weshaggard) [inline eng/common/scripts/modules/ChangeLog-Operations.psm1:2]: Is this for testing or is this a change you are planning to add?

File: eng/pipelines/eng-common-sync.yml [hunk 5]
```diff
@@ -30,12 +30,25 @@ stages:
         displayName: Sync ${{ parameters.DirectoryToSync }} Directory
 
         steps:
+        - ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
+          - pwsh: |
+              $patchDir = "$(System.DefaultWorkingDirectory)/patchfiles"
+              $patchfiles = git format-patch HEAD~1...origin/master -- ${{ parameters.DirectoryToSync }} -o $patchDir
```
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:36]:
```suggestion
              $patchfiles = git format-patch HEAD...master -- ${{ parameters.DirectoryToSync }} -o $patchDir
```

Why did you need to back-up one commit?

File: eng/pipelines/eng-common-sync.yml [hunk 6]
```diff
@@ -30,12 +30,25 @@ stages:
...
+              if ($patchfiles)
```
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:37]: Is this the correct condition? What is the output of the format-patch command? I would expect it to have output for successes and failures. Perhaps we need to use should use the exit code?

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 7]
```diff
@@ -17,12 +17,23 @@ steps:
...
-
-        if (!(Test-Path $repoPath)) { mkdir $repoPath }
-        Remove-Item -v -r $repoPath
-        Copy-Item -v -r $(Build.SourcesDirectory)/${{ parameters.DirectoryToSync }} $repoPath
-        Get-ChildItem -r $repoPath
-
+        if ($PatchFilesLocation)
+        {
+          pushd ${{ repo }}
+          $results = Get-ChildItem $PatchFilesLocation | % { git am $_.FullName }
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:23]: What ends up in the $results array? Wouldn't it be all the outputted lines?

File: eng/common/pipelines/templates/steps/create-pull-request.yml [hunk 8]
```diff
@@ -37,6 +37,7 @@ steps:
     }
 
   displayName: Check for changes
+  condition: and(succeeded(), eq(variables['HasChanges'], ''))
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/create-pull-request.yml:40]: While this change might actually work for our new scenario it will potentially break scenarios people use this template multiple times in the same pipeline. Any usage after the first will always get the result of the first execution of this step which would be incorrect. We will need to make a tweak to how we do this. Instead I think we should create a new template parameter called "SkipCheckingForChanges" which we can use to exclude this step if it is set.

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 9]
```diff
@@ -17,12 +17,23 @@ steps:
...
+          if ($results.Count -ne (Get-ChildItem $PatchFilesLocation).Count)
+          {
+            Write-Output "Failed to apply patch files"
+            exit 1
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:27]: We should set the Has Changes variable in this step instead of the step that produces the patches as the changes really depends on what is already in each of the language repos.

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 10]
```diff
@@ -17,12 +17,23 @@ steps:
     - pwsh: |
         git clone --branch ${{ parameters.BaseBranchName }} https://github.com/azure/${{ repo }}
         $repoPath = "${{ repo }}/${{ parameters.DirectoryToSync }}"
-
-        if (!(Test-Path $repoPath)) { mkdir $repoPath }
-        Remove-Item -v -r $repoPath
-        Copy-Item -v -r $(Build.SourcesDirectory)/${{ parameters.DirectoryToSync }} $repoPath
-        Get-ChildItem -r $repoPath
-
+        if ($PatchFilesLocation)
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:20]: Lookin
```suggestion
        if ($(PatchFilesLocation))
```

File: eng/pipelines/templates/steps/sync-directory.yml (hunk 7 above)
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:23]:
```suggestion
          $results = Get-ChildItem $(PatchFilesLocation) | % { git am $_.FullName }
```

Comment (chidozieononiwu) [reply eng/pipelines/templates/steps/sync-directory.yml:23]: This git am works correctly it $result will contain a line showing information for each applied patch file which should equal the number of patch files. If the number of lines in result don't match the number of patch files then an error occurred somewhere.

File: eng/pipelines/eng-common-sync.yml (hunk 6 above)
Comment (chidozieononiwu) [reply eng/pipelines/eng-common-sync.yml:37]: The output is the list of all the patch files for success. Failure situations will write to console but not output anything

File: eng/pipelines/eng-common-sync.yml (hunk 5 above)
Comment (chidozieononiwu) [reply eng/pipelines/eng-common-sync.yml:36]: `$(System.PullRequest.SourceBranch)` is actually not present because as you pointed out it is in a detached HEAD state, with HEAD being the pull request merge commit. So backing up one commit is basically going one commit before the merge commit which is what we need.

File: eng/common/scripts/modules/ChangeLog-Operations.psm1 (hunk 4 above)
Comment (chidozieononiwu) [reply eng/common/scripts/modules/ChangeLog-Operations.psm1:2]: Yes this is for testing. I will clean it up before merging.

File: eng/pipelines/eng-common-sync.yml (hunk 6 above)
Comment (weshaggard) [reply eng/pipelines/eng-common-sync.yml:37]: >Failure situations will write to console but not output anything

I'm not sure I understand this. it does output things to the console which is what would go into $patchFiles variable correct?

File: eng/pipelines/templates/steps/sync-directory.yml (hunk 7 above)
Comment (weshaggard) [reply eng/pipelines/templates/steps/sync-directory.yml:23]: We should try to use exit codes if they are correctly set as those are a little better then depending on the console output (especially the number of lines in the output).

File: eng/pipelines/eng-common-sync.yml (hunk 5 above)
Comment (weshaggard) [reply eng/pipelines/eng-common-sync.yml:36]: OK you are correct the sourcebranch does not appear in the main repo when the PR is from a fork. So perhaps we should use `refs/remotes/pull/995/merge` or HEAD. I don't think we need to move back one commit.

Comment (weshaggard) [reply eng/pipelines/eng-common-sync.yml:36]: However we should still use the TargetBranch

File: eng/pipelines/eng-common-sync.yml (hunk 6 above)
Comment (chidozieononiwu) [reply eng/pipelines/eng-common-sync.yml:37]: From my testing when it fails it write an info about why it fails to console but the `$patchfiles`  variable remains `$null`.

File: eng/pipelines/eng-common-sync.yml (hunk 5 above)
Comment (chidozieononiwu) [reply eng/pipelines/eng-common-sync.yml]: `$(System.PullRequest.TargetBranch)` is just master which does not work as a revision. It needs to be `origin/master`

File: eng/pipelines/templates/steps/sync-directory.yml (hunk 7 above)
Comment (chidozieononiwu) [reply eng/pipelines/templates/steps/sync-directory.yml:23]: There is still scenario where the $LASTEXITCODE is 0 but no patch file was created. I will make use of both.

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 11]
```diff
@@ -15,14 +16,28 @@ steps:
...
+        {
+          pushd ${{ repo }}
+          $results = Get-ChildItem $(PatchFilesLocation) | % { git -c user.name="azure-sdk" -c user.email="azuresdk@microsoft.com" am $_.FullName }
+          if (($results.Count -ne (Get-ChildItem $(PatchFilesLocation)).Count) -or ($LASTEXITCODE -ne 0))
+          {
+            Write-Output "Failed to apply patch files to [https://github.com/azure/${{ repo }}]"
+            exit 1
+          }
+          echo "##vso[task.setvariable variable=HasChanges]$true"
+          echo "##vso[task.setvariable variable=RepoNameWithoutOwner]$repo"
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:32]: This is a side-effect of me putting the RepoNameWithoutOwner as part of the same step as the check for changes. We should probably separate that out so we don't have to do this here.

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 12]
```diff
@@ -15,14 +16,28 @@ steps:
...
+        Write-Host "Patch 1 $(PatchFilesLocation)"
         git clone --branch ${{ parameters.BaseBranchName }} https://github.com/azure/${{ repo }}
         $repoPath = "${{ repo }}/${{ parameters.DirectoryToSync }}"
-
-        if (!(Test-Path $repoPath)) { mkdir $repoPath }
-        Remove-Item -v -r $repoPath
-        Copy-Item -v -r $(Build.SourcesDirectory)/${{ parameters.DirectoryToSync }} $repoPath
-        Get-ChildItem -r $repoPath
-
+        if ('$(PatchFilesLocation)')
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:22]: Should this be a file exists check?

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 13]
```diff
@@ -15,14 +16,28 @@ steps:
...
+        {
+          pushd ${{ repo }}
+          $results = Get-ChildItem $(PatchFilesLocation) | % { git -c user.name="azure-sdk" -c user.email="azuresdk@microsoft.com" am $_.FullName }
+          if (($results.Count -ne (Get-ChildItem $(PatchFilesLocation)).Count) -or ($LASTEXITCODE -ne 0))
+          {
+            Write-Output "Failed to apply patch files to [https://github.com/azure/${{ repo }}]"
+            exit 1
+          }
+          echo "##vso[task.setvariable variable=HasChanges]$true"
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:31]: We need to write to reverse HasChanges=false if there are no changes.

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 14]
```diff
@@ -15,14 +16,28 @@ steps:
...
-        if (!(Test-Path $repoPath)) { mkdir $repoPath }
-        Remove-Item -v -r $repoPath
-        Copy-Item -v -r $(Build.SourcesDirectory)/${{ parameters.DirectoryToSync }} $repoPath
-        Get-ChildItem -r $repoPath
-
+        if ('$(PatchFilesLocation)')
+        {
+          pushd ${{ repo }}
+          $results = Get-ChildItem $(PatchFilesLocation) | % { git -c user.name="azure-sdk" -c user.email="azuresdk@microsoft.com" am $_.FullName }
+          if (($results.Count -ne (Get-ChildItem $(PatchFilesLocation)).Count) -or ($LASTEXITCODE -ne 0))
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:26]: Shouldn't we need to check $lastExitcode after each call to git am?

File: eng/pipelines/eng-common-sync.yml [hunk 15]
```diff
@@ -30,16 +30,38 @@ stages:
         displayName: Sync ${{ parameters.DirectoryToSync }} Directory
 
         steps:
+        - ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
+          - pwsh: |
+              $patchDir = "$(Build.ArtifactStagingDirectory)/patchfiles"
+              New-Item $patchDir -Type Directory
+              $patchfiles = git format-patch HEAD...origin/master -- "${{ parameters.DirectoryToSync }}"
+              Copy-Item -Path "$(Build.SourcesDirectory)/*.patch" -Destination $patchDir
```
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:38]: We should be a able to pass `--output-directory` directly to the format-patch command.

File: eng/pipelines/eng-common-sync.yml [hunk 16]
```diff
@@ -30,16 +30,38 @@ stages:
         displayName: Sync ${{ parameters.DirectoryToSync }} Directory
 
         steps:
+        - ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
+          - pwsh: |
+              $patchDir = "$(Build.ArtifactStagingDirectory)/patchfiles"
+              New-Item $patchDir -Type Directory
+              $patchfiles = git format-patch HEAD...origin/master -- "${{ parameters.DirectoryToSync }}"
```
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:37]:
```suggestion
              $patchfiles = git format-patch HEAD...$(system.pullRequest.targetBranch)  -- "${{ parameters.DirectoryToSync }}"
```
This needs to be target branch in order to match the other logic in this file and to allow us to support other branches.

File: eng/pipelines/eng-common-sync.yml [hunk 17]
```diff
@@ -30,16 +30,38 @@ stages:
         displayName: Sync ${{ parameters.DirectoryToSync }} Directory
 
         steps:
+        - ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
+          - pwsh: |
+              $patchDir = "$(Build.ArtifactStagingDirectory)/patchfiles"
```
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:35]: To help with diagnosing issues it might be worth publishing these patchfiles as an artifact so we can download and try them if something goes wrong.

File: eng/pipelines/eng-common-sync.yml [hunk 18]
```diff
@@ -30,16 +30,38 @@ stages:
         displayName: Sync ${{ parameters.DirectoryToSync }} Directory
 
         steps:
+        - ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
```
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:33]: We should consider setting `Set-PsDebug -Trace 1` like https://github.com/Azure/azure-sdk-tools/blob/master/eng/pipelines/templates/steps/sync-repo-branch.yml#L25 for these larger inline pwsh scripts so that can more easily diagnose issues. This tracing will output every line of the script as it executes it.

File: eng/pipelines/eng-common-sync.yml (hunk 15 above)
Comment (chidozieononiwu) [reply eng/pipelines/eng-common-sync.yml:38]: Strangely that command did not work. I tried it both on local and on the remote server but it wont place the patch files in the specified directory. There might be an issue with git, or am just not doing it properly.

Comment (weshaggard) [reply eng/pipelines/eng-common-sync.yml]: I used `-o` successfully locally.

File: eng/pipelines/templates/steps/sync-directory.yml (hunk 12 above)
Comment (chidozieononiwu) [reply eng/pipelines/templates/steps/sync-directory.yml:22]: The variable is only set if patch files were created so just checking if it exists is sufficient.

File: eng/pipelines/eng-common-sync.yml (hunk 15 above)
Comment (chidozieononiwu) [reply eng/pipelines/eng-common-sync.yml]: Yes that only seems to work if the output directory is within the source. I was trying to out then in the Artifact Staging Directory.

Comment (weshaggard) [reply eng/pipelines/eng-common-sync.yml:39]: If we cannot get the output directory to work then we should do a move instead of a copy so that we leave the repo in a clean state.

File: eng/pipelines/eng-common-sync.yml [hunk 19]
```diff
@@ -30,16 +30,46 @@ stages:
...
+              }
+              else {
+                Write-Host "Failed to Create PatchFiles from Pull Request [https://github.com/Azure/azure-sdk-tools/pull/$(System.PullRequest.PullRequestNumber)]"
+                exit 1
+              }
+            displayName: Create Patch Files from Changes in PR
+            workingDirectory: $(System.DefaultWorkingDirectory)
+
+          - task: PublishPipelineArtifact@1
+            condition: succeeded()
```
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:56]: Should we condition this on whether or not PatchFilesLocation is set.

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 20]
```diff
@@ -15,14 +16,31 @@ steps:
 
   - ${{ each repo in parameters.Repos }}:
     - pwsh: |
+        Set-PsDebug -Trace 1
+        Write-Host "Patch 1 $(PatchFilesLocation)"
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:20]: Might want to remove this write-host.

File: eng/pipelines/templates/steps/sync-directory.yml (hunk 12 above)
Comment (weshaggard) [reply eng/pipelines/templates/steps/sync-directory.yml:23]: I guess we can do with just an exists check instead of a file exists

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 21]
```diff
@@ -15,14 +16,31 @@ steps:
...
-        Remove-Item -v -r $repoPath
-        Copy-Item -v -r $(Build.SourcesDirectory)/${{ parameters.DirectoryToSync }} $repoPath
-        Get-ChildItem -r $repoPath
-
+        if ('$(PatchFilesLocation)')
+        {
+          pushd ${{ repo }}
+          foreach ($file in (Get-ChildItem $(PatchFilesLocation)))
+          {
+            $result = git -c user.name="azure-sdk" -c user.email="azuresdk@microsoft.com" am $file.FullName
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:28]: What happens if there is not this patch doesn't apply? We don't want it to fail in that case.

Comment (chidozieononiwu) [reply eng/pipelines/templates/steps/sync-directory.yml:27]: I updated it to not exit but set HasChanges to $false

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 22]
```diff
@@ -15,14 +16,35 @@ steps:
...
-        Get-ChildItem -r $repoPath
-
+        if ('$(PatchFilesLocation)')
+        {
+          pushd ${{ repo }}
+          $results = @()
+          echo "##vso[task.setvariable variable=HasChanges]$false"
+          foreach ($file in (Get-ChildItem $(PatchFilesLocation)))
+          {
+            $results += git -c user.name="azure-sdk" -c user.email="azuresdk@microsoft.com" am $file.FullName
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:29]: Shouldn't you still check the lastExitCode after each invoke of git am?

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 23]
```diff
@@ -15,14 +16,35 @@ steps:
...
+          }
+          if (($results.Count -eq (Get-ChildItem $(PatchFilesLocation))) -and ($LASTEXITCODE -eq 0))
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:31]: `(Get-ChildItem $(PatchFilesLocation))` that doesn't return a count does it?

File: eng/pipelines/eng-common-sync.yml [hunk 24]
```diff
@@ -30,16 +30,42 @@ stages:
         displayName: Sync ${{ parameters.DirectoryToSync }} Directory
 
         steps:
+        - ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
+          - pwsh: |
+              Set-PsDebug -Trace 1
+              $patchDir = "$(Build.ArtifactStagingDirectory)/patchfiles"
+              New-Item $patchDir -Type Directory
+              $patchfiles = git format-patch HEAD...origin/$(system.pullRequest.targetBranch) -- "${{ parameters.DirectoryToSync }}"
```
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:38]: From my local testing you don't need to create the patch directory nor do you have to move them. You can definitely use the --output-directory. Here is the command that I was able to use.

```
$patchfiles = git format-patch --output-directory $patcheDir HEAD...origin/$(system.pullRequest.targetBranch) -- "${{ parameters.DirectoryToSync }}"
```

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 25]
```diff
@@ -15,14 +16,35 @@ steps:
...
-
+        if ('$(PatchFilesLocation)')
+        {
+          pushd ${{ repo }}
+          $results = @()
+          echo "##vso[task.setvariable variable=HasChanges]$false"
+          foreach ($file in (Get-ChildItem $(PatchFilesLocation)))
+          {
+            $results += git -c user.name="azure-sdk" -c user.email="azuresdk@microsoft.com" am $file.FullName
+          }
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:30]:
```suggestion
         $shaBeforePatches = git ref-parse --short HEAD
          foreach ($file in (Get-ChildItem $(PatchFilesLocation)*.patch))
          {
             git -c user.name="azure-sdk" -c user.email="azuresdk@microsoft.com" am -3 $file.FullName
             if ($lastExitCode -ne 0) { 
               git am --show-current-patch=diff
               
               # Once we hit one conflict that causes the patch to fail we should fail the pipeline until that
               # gets addressed. 
               Write-Error "##vso[task.LogIssue type=warning;]Failed to properly apply patch files to [https://github.com/azure/${{ repo }}]"
               exit 1
             }
          }
          $shaAfterPatches = git ref-parse --short HEAD
          $hasChanges = $shaBeforePatches -ne $shaAfterPatches
          echo "##vso[task.setvariable variable=HasChanges]$hasChanges"
```

**Comment by chidozieononiwu** [comment]
Am closing this since the changes have been incorporated in https://github.com/Azure/azure-sdk-tools/pull/1052

**Comment by weshaggard** [comment]
@chidozieononiwu please also go and close the sync PRs.
